input-output-hk/cardano-sl
# [CBR-289] Review mnemonic implementation

## Description

This is actually an implementation of BIP-39. The current implementation is inspired from Haskoin and is rather hard to grasp. This refactor merges together `Pos.Util.BackupPhrase` and `Pos.Util.Mnemonic`. The entire implementation has been reviewed to rely on smart constructors and maintain invariant this way. 

Quite some tests have been added, and I am open to any idea of things that could also be tested and that aren't currently included. 

> Note
> At the moment, the PR still contains the old code, moved in a test folder and used for comparison 
> with Quickcheck. This is to make sure that the new code behaves exactly as the old one
> since it touches to a quite sensitive part of the code. I am not sure if this has to stay or can be removed if we all get convinced that the behavior has been preserved (keeping only the relevant tests).

## Linked issue

[[CBR-289]](https://iohk.myjetbrains.com/youtrack/issue/CBR-289)

## Type of change


## Developer checklist


## Testing checklist


## QA Steps

- We need to make sure that once this get merged, users with an existing backup phrase can still restore their wallet (and it has to be the same wallet). Said differently, changing the implementation shouldn't have changed the behavior.

- It shouldn't be possible to use the example mnemonic advertised in the API doc anymore

- Paper wallet already emitted should also not be affected and we must verify they can still be redeemed as expected.

## Screenshots (if available)
